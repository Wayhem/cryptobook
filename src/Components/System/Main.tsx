import { ThemeProvider } from 'styled-components'
import { Provider } from 'react-redux'
import { store } from 'Store'
import { GlobalStyle, theme } from 'Config/styled'

const App = (): JSX.Element => (
  <>
    <GlobalStyle />
    <Provider store={store}>
      <ThemeProvider theme={theme}>hello</ThemeProvider>
    </Provider>
  </>
)

export default App
