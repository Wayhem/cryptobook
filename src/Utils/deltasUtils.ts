import { XBTUSD_GROUPS_ENUM, ETHUSD_GROUPS_ENUM } from 'Models/constants/groups'
import Delta, { DeltaWithoutTotal } from 'Models/Delta'
import OrdersBook from 'Models/OrderBook'

function getClosestMultiple(x: number, y: number) {
  return Math.round(x / y) / (1 / y)
}

function isMultiple(x: number, y: number) {
  return getClosestMultiple(x, y) === x
}

const groupDeltasByNumber = (group: XBTUSD_GROUPS_ENUM | ETHUSD_GROUPS_ENUM, deltas: Delta[]) => {
  const groupedDeltas = [] as Delta[]
  if (Array.isArray(deltas) && deltas.length) {
    deltas.forEach(delta => {
      if (groupedDeltas.length === 0) {
        groupedDeltas.push([getClosestMultiple(delta[0], group), delta[1], delta[2]])
      } else if (isMultiple(delta[0], group) && delta[0] !== groupedDeltas[groupedDeltas.length - 1][0]) {
        groupedDeltas.push(delta)
      } else if (getClosestMultiple(delta[0], group) === groupedDeltas[groupedDeltas.length - 1][0]) {
        groupedDeltas[groupedDeltas.length - 1][1] += delta[1]
        groupedDeltas[groupedDeltas.length - 1][2] += delta[2]
      } else {
        groupedDeltas.push([getClosestMultiple(delta[0], group), delta[1], delta[2]])
      }
    })
  }
  return groupedDeltas
}

const getUpdatedDeltas = (newDeltas: DeltaWithoutTotal[], currentOrderBook: OrdersBook, isAsks: boolean): Delta[] => {
  const currentDeltas = isAsks ? [...currentOrderBook.asks] : [...currentOrderBook.bids]

  newDeltas.forEach((newDelta: DeltaWithoutTotal) => {
    if (newDelta[1] === 0) {
      for (const [currentDeltaIndex, currentDelta] of currentDeltas.entries()) {
        if (currentDelta[0] === newDelta[0]) {
          currentDeltas.splice(currentDeltaIndex, 1)
          break
        }
      }
    } else if (
      isAsks
        ? newDelta[0] > currentDeltas[currentDeltas.length - 1][0]
        : newDelta[0] < currentDeltas[currentDeltas.length - 1][0]
    ) {
      currentDeltas.push([newDelta[0], newDelta[1], newDelta[1]])
    } else if (isAsks ? newDelta[0] < currentDeltas[0][0] : newDelta[0] > currentDeltas[0][0]) {
      currentDeltas.unshift([newDelta[0], newDelta[1], newDelta[1]])
    } else {
      for (const [currentDeltaIndex, currentDelta] of currentDeltas.entries()) {
        if (currentDelta[0] === newDelta[0]) {
          currentDeltas[currentDeltaIndex][1] = newDelta[1]
          currentDeltas[currentDeltaIndex][2] += newDelta[1]
        } else if (
          isAsks
            ? currentDelta[0] > newDelta[0] && currentDeltas[currentDeltaIndex - 1][0] < newDelta[0]
            : currentDelta[0] < newDelta[0] && currentDeltas[currentDeltaIndex - 1][0] > newDelta[0]
        ) {
          currentDeltas.splice(currentDeltaIndex, 0, [newDelta[0], newDelta[1], newDelta[1]])
        }
      }
    }
  })

  return currentDeltas
}

export { groupDeltasByNumber, getUpdatedDeltas }
