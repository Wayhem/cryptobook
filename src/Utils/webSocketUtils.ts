import WSMessage from 'Models/WSMessage'
import ProductIds from 'Models/constants/productIds'
import WSMessageEvents from 'Models/constants/webSocketMessageEvents'
import FeedTypes from 'Models/constants/feedTypes'
import { XBTUSD_GROUPS_ENUM } from 'Models/constants/groups'
import Delta from 'Models/Delta'

const buildToggleWSMessage = (event: WSMessageEvents, productId: ProductIds): WSMessage => ({
  event,
  feed: FeedTypes.standard,
  product_ids: [productId],
})

function getClosestMultiple(x: number, y: number) {
  return Math.round(x / y) / (1 / y)
}

function isMultiple(x: number, y: number) {
  return getClosestMultiple(x, y) === x
}

const groupDeltasByNumber = (group: XBTUSD_GROUPS_ENUM, deltas: Delta[]) => {
  const groupedDeltas = [] as Delta[]
  if (Array.isArray(deltas) && deltas.length) {
    deltas.forEach(delta => {
      if (groupedDeltas.length === 0) {
        groupedDeltas.push([getClosestMultiple(delta[0], group), delta[1], delta[2]])
      } else if (isMultiple(delta[0], group) && delta[0] !== groupedDeltas[groupedDeltas.length - 1][0]) {
        groupedDeltas.push(delta)
      } else if (getClosestMultiple(delta[0], group) === groupedDeltas[groupedDeltas.length - 1][0]) {
        groupedDeltas[groupedDeltas.length - 1][1] += delta[1]
        groupedDeltas[groupedDeltas.length - 1][2] += delta[2]
      } else {
        groupedDeltas.push([getClosestMultiple(delta[0], group), delta[1], delta[2]])
      }
    })
  }
  return groupedDeltas
}

export { buildToggleWSMessage, groupDeltasByNumber }
