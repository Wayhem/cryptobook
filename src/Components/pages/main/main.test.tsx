import Main from 'Components/pages/main'
import { render } from 'Utils/test-utils'

describe('Main page component', () => {
  test('Main renders and matches snapshot', () => {
    const { container } = render(<Main />)
    expect(container).toMatchSnapshot()
  })
})
