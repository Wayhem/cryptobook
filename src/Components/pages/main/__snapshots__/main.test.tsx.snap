// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Main page component Main renders and matches snapshot 1`] = `
<div>
  <div
    class="sc-kEjbQP gyJgOz"
  >
    <div
      class="sc-iqHYmW dczUNV"
    >
      <div
        class="sc-bdfBQB dycULH"
      >
        <div
          class="sc-gsTEea dRwGRi"
        >
          Order Book
        </div>
        <div>
          <label
            class="sc-hKgJUU TebGZ"
          >
            group:
          </label>
          <select
            class="sc-dlfnuX gNzwCX"
          >
            <option
              value="0.5"
            >
              0.5
            </option>
            <option
              value="1"
            >
              1
            </option>
            <option
              value="2.5"
            >
              2.5
            </option>
          </select>
        </div>
      </div>
      <div
        class="sc-crrszt bRAvch"
      >
        <div
          class="sc-pGacB eBMqEt"
        >
          <div
            class="sc-jSgvzq gacCAu"
          >
            <div
              class="sc-iBPTik tGfiW"
            >
              total
            </div>
            <div
              class="sc-iBPTik tGfiW"
            >
              size
            </div>
            <div
              class="sc-iBPTik tGfiW"
            >
              price
            </div>
          </div>
        </div>
        <div
          class="sc-pGacB eBMqEt"
        >
          <div
            class="sc-jSgvzq gacCAu"
          >
            <div
              class="sc-iBPTik tGfiW"
            >
              price
            </div>
            <div
              class="sc-iBPTik tGfiW"
            >
              size
            </div>
            <div
              class="sc-iBPTik tGfiW"
            >
              total
            </div>
          </div>
        </div>
      </div>
      <div
        class="sc-dQoVA dCRteQ"
      >
        <button
          class="sc-eCstlR gqljiU"
          disabled=""
        >
          Toggle feed
        </button>
        <button
          class="sc-eCstlR cviYRt"
        >
          Start feed
        </button>
      </div>
    </div>
  </div>
</div>
`;
